### PR TITLE
Add a debug log.

### PR DESCRIPTION
**Pull request by cowlicks**
This stores messages passed to the `log` function. 

I think this might be useful to help debug user's issues that we can't reproduce. Or we could ask users to include it in their bug reports.

If we decide to more vocally ask user to post debugging info with their issues (like with an `ISSUE_TEMPLATE.md`)  then the debug log should be exposed through the option menu, along with some basic redaction to hide URL's the user has visited.

The log messages are stored in a ring buffer (`DebugLog.logBook`), and adding an entry is O(1).

### CODE REVIEW

**Comment by cowlicks** [comment]
@ghostwords I don't think we are using `log` enough in our codebase to make this immediately useful. But once we have more log statements it would allow users to include more helpful information when they report issues. 

This loosely ties into a plan I have for improving logging and the way we record tracking. This should make diagnosing site breakages easier, and eventually it would let us improve our blocking behavior. It would also allow us to better understand what kind of tracking is happening out there on the internet.

**Comment by ghostwords** [comment]
Cool, makes sense. This pull request is about efficiently capturing log messages. Making logging more helpful and sensible is orthogonal to this.

How much extra memory use does capturing 1,000 typical Privacy Badger messages entail? Is worst case (only the longest message?) significantly worse?

**Comment by ghostwords** [comment]
Related : #1110

**Comment by cowlicks** [comment]
I've recently been wanting this. I'll pick it back up soon.

**Comment by cowlicks** [comment]
The usecase here is when someone reports a bug, they can also post this log, which be helpful.

A user can get the log by running `JSON.stringify(badger.debugLog.output(), null, 2)` in the background context.

**Comment by cowlicks** [comment]
@ghostwords I rebased this and reduced the debug log size from 1000 entries to 300. Ready for review.

**Comment by ghostwords** [comment]
My current thinking is that permanent (as opposed to temporary, while developing) logging in Privacy Badger isn't helpful for debugging, and it isn't clear to me it's worth making it helpful. It's just my impression from fielding user support requests so far. We often do more need more information from users, but most often that information is the contents of snitch/action maps. I currently ask users to run a snippet of code (which is somewhat but not much more annoying, I think, than having a [dedicated debug function](1430)), with the next step being just [making that information available in the UI](1289).

On the other hand, you seem very interested in giving this a shot; the ring buffer here is mostly orthogonal to the rest of Privacy Badger, with some impact on memory and (probably tiny impact on) CPU use. We should confirm though ... We only console.log when DEBUG is enabled, but we always write to this buffer for all users. Have you checked how much extra memory a full buffer imposes? Would we have an issue if we added a long, frequently occurring `log` statement (maybe one that's dynamically built from request data)?

**Comment by cowlicks** [comment]
@ghostwords part of my intention with this and #1430 is to reduce the amount of friction with users's who report issues on github. It removes the first round trip of us asking, "hey can you go run this code?", then waiting for them to report back (potentially reporting back that they were unable to reproduce their own bug).

It should give us some perspective on the state the badger is in, which will let us triage issues more quickly bugs faster to triage. I agree that our current distribution of log statements in the codebase could be improved, but we can't really improve that until we start looking at it.

If this proves unnecessary, it can be removed. If it is helpful I think the next step would be adding `log` statements that would identify common issues, and eventually exposing a ui.

**Comment by cowlicks** [comment]
I'll profile this, and add some sample output soon.

**Comment by cowlicks** [comment]
I also think this would be useful for people who send us support emails. I'll ask Amul.

**Comment by amullionaire** [comment]
Hi! I think this would be great if it eliminates the step where we have to ask less questions from people to figure out how we can help. For instance, what other extensions they're running, which version of PB & browser they have etc.

**Comment by cowlicks** [comment]
FYI while I have not measure the performance impact of this I expect it to be small. It is currently ready for review despite not including performance numbers.

One thing has been introduced to the codebase since I wrote this. We included a switch on a debug statement in https://github.com/EFForg/privacybadger/pull/1482/commits/ab741772020f1323867a6f2c80a7d4504165d5b0

I'll remove this before merging.

**Comment by cowlicks** [comment]
I've found this especially useful for collecting internal privacy badger data while conducting crawls.

**Comment by cowlicks** [comment]
I did 2 crawls, of 25 sites. This was enough to fill the debug log several times over.
* on master: [`Browser 71 MB private, 111 MB shared`](https://github.com/cowlicks/privacybadgerchrome/blob/pr-1221-master-crawl/crawls/master.35223a3af0794bd6ffc73a319558130654818e79.results.json#L1811)
* on this branch [`Browser 69 MB private, 109 MB shared`](https://github.com/cowlicks/privacybadgerchrome/blob/pr-1221-crawl/debug-log.56e2de89f4c44059b105061e40cb0e31f6452070.results.json#L2989)

Basically the memory usage is within the error margin of this method. Consecutive crawls vary by ~5 MB private and ~10 MB shared. So this PR does not have a significant impact on memory usage.

**Comment by ghostwords** [comment]
This PR does not remove the step of asking users for more information.

While this PR provides potentially useful groundwork for the future, I am not aware of running into any issues this would have helped solve.

We may come back to this in the future, but in the meantime we should focus on more clearly and immediately useful enhancements.
